File: Programming Languages/JavaScript/0049_Group_Anagrams/solution_2.js
// 0049. Group Anagrams
// Difficulty : Medium
// Tags       : array, hash-table, string, sorting
// Solved on  : 2026-05-18
// Attempt    : #2
// LeetCode   : https://leetcode.com/problems/group-anagrams/

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    var map = new Map();
    strs.forEach((str)=>{
        const count = new Array(26).fill(0);
        for(const x of str){
            const ascii = x.codePointAt(0) -'a'.codePointAt(0);
            count[ascii]++;
        }
        const key = count.join("#"); // with character counting array makes a unique string key
        if(!map.has(key)){
            map.set(key,[]);
        }
        map.get(key).push(str);
    });
    return [...map.values()]
    
};

File: DSA/Array/0049_Group_Anagrams/python3/solution_2.py
# 0049. Group Anagrams
# Difficulty : Medium
# Tags       : array, hash-table, string, sorting
# Solved on  : 2026-05-18
# Attempt    : #2
# LeetCode   : https://leetcode.com/problems/group-anagrams/

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        mp = {}
        for _,x in enumerate(strs):
            count = [0] * 26
            for _,ch in enumerate(x):
                asc = ord(ch) - ord('a')
                count[asc]+=1
            key = ",".join(map(str,count))
            if key not in mp:
                mp[key] = []
            mp[key].append(x)
        return [val for _,val in enumerate(mp.values())]
        